### DISCUSSION TITLE
Wifi connection not working on newer ESP32 WROOM boards

### DISCUSSION THREAD

**Discussion by gkzsolt** [micropython/micropython#10521] [ESP32] [answered]
This is a duplicate of #10377. I am sorry, but cannot help it to be somehow desperate about Micropython not working on the newer WROOM revision 3 dev boards. In a nutshell, network.WLAN().connect works on ESP32D0WDQ6 (revision 1) boards, but not on ESP32D0WDQ5 (revision 3), where a board reset occurs.
My electronics provider says that:
> "Q5" és a "Q6" refer to the IC packaging. Between revision 1 and 3 the latter is the newer, where the manufacturer issued several fixes, which could lead to the errors encountered. 

They provided a link to the [ESP32 Series SoC Errata](https://www.espressif.com/sites/default/files/documentation/esp32_errata_en.pdf), which did not help much. It might be useful to MicroPython developers though. Is any developer following this?

I tried all versions available at the download page, with firmware compiled with ESP-IDF v3.x or v4.x. Up to v1.11 the code just block without returning; higher versions all were resetting the board (tried also nightly builds: e.g. v1.19.1-812-g67097d8c2 (2023-01-18) .bin)

Unfortunately - or well, fortunately - the C/C++ compiled code from Arduino IDE works well, so this is a MicroPython issue.

I would prefer the board with the older chip to use in my projects, but the distributor does not have those. And when you order a board from a distributor, you cannot know the chip version used.

Is there any hope for this to be considered/tackled in the future? What can I do?

Thank you,

**Comment by robert-hh** [comment, marked as answer]
I have a board here with a WROOM module, which reports to `esptool.py chip_id` with `Chip is ESP32-D0WD-V3 (revision 3)`. And that one works well with micropython. What does you device report on `esptool.py chip_id`?

**Reply by gkzsolt** [reply]
```
$ esptool.py chip_id
esptool.py v2.8
Found 1 serial ports
Serial port /dev/ttyUSB0
Connecting....
Detecting chip type... ESP32
Chip is ESP32D0WDQ5 (revision 3)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 40:22:d8:7b:35:c0
Uploading stub...
Running stub...
Stub running...
Warning: ESP32 has no Chip ID. Reading MAC instead.
MAC: 40:22:d8:7b:35:c0
```

**Reply by robert-hh** [reply]
The errata you referenced and the data sheet does not mention ESP32D0WDQ5.  I found a hint at another place:
> There is indeed an electrical difference between these 2 versions of the chip. In IDE Arduino config, the V6 works with Flash Mode = "DIO" and not with Flash Mode = "QIO". The V5 works with Flash QIO mode.

**Reply by robert-hh** [reply]
On a second thought, a wrong flash mode should stop the chip from working at all.

**Reply by gkzsolt** [reply]
Thanks a lot. But how can I use this information with Micropython?

**Reply by robert-hh** [reply]
I have to revert my message: The V3 board I have fails with WiFi as well. I never tried WiFi with these, just Ethernet. Time to start debugging.

**Reply by gkzsolt** [reply]
It flashed OK, I could connect through the USB UART and flash a LED. So basic things are working, but not the wifi network.

**Reply by robert-hh** [reply]
I changed the discussion into an issue. Next steps tomorrow.

**Reply by robert-hh** [reply]
I recalled a problem noted for S3 devices, and that's the issue here as well.
ifconfig() must not be called before the connection is established. I had that in my script. So please check your connection script, whether ifconfig() is called and whther that happens before the call to connect().

**Reply by davefes** [reply]
> ifconfig() must not be called before the connection is established

Sounds like one of the rules I have been trying to find for a long time.  Setting static IPs has been the biggest headache for me.  Is there a reference document or did you learn the hard-way?

**Reply by robert-hh** [reply]
> a problem noted for S3 devices

Wrong recall. That problem exists even for my oldest ESP32 boards. It seems more an issue of the changed esp-idf. May be not related to the discussion here.

**Reply by robert-hh** [reply]
> Sounds like one of the rules I have been trying to find for a long time. 

For a long time I used to set the static IP address **before** calling connect(). But since recently that fails.

**Reply by davefes** [reply]
Over the last 6 months or so setting a static IP before connect has been working for me on V1.19.1 generic and SPIRAM.

Thanks anyhow.

**Reply by gkzsolt** [reply]
> that's the issue here as well. ifconfig() must not be called before the connection is established. I had that in my script. So please check your connection script, whether ifconfig() is called and whther that happens before the call to connect().

Please read my original post (#10377) again. What I did was only what is in the doc, and in that order:

The following code:
```
import network
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
```
runs, than
```
wlan.connect('your-ssid', 'your-key')  # or wlan.scan()
```
resets the board:
```
ets Jul 29 2019 12:21:46

rst:0x8 (TG1WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
...
W (61) boot.esp32: PRO CPU has been reset by WDT.
W (61) boot.esp32: WDT reset info: PRO CPU PC=0x40093e52
W (63) boot.esp32: WDT reset info: APP CPU PC=0x4008d3a2
```
There is no `ifconfig`, or setting static IP address here.

**Reply by robert-hh** [reply]
@gkzsolt  That's a right way of connecting, but also a pity, because then I cannot reproduce the error. That sequence works here with my V3 board.

**Reply by robert-hh** [reply]
Could you try to add a short delay after `wlan.active(True)`, like `time.sleep_ms(10)`. Unless you enter the code lines anyhow one by one form REPL.

**Reply by gkzsolt** [reply]
> Could you try to add a short delay after `wlan.active(True)`, like `time.sleep_ms(10)`. Unless you enter the code lines anyhow one by one form REPL.

I did that: entered the code line by line from REPL.

**Reply by robert-hh** [reply]
Which board exactly do you use. Can you give a link to that board? NodeMCU ESP32 is kind of generic with many vendors selling boards under that name.

**Reply by gkzsolt** [reply]
It is [this](https://www.hestore.hu/prod_10037961.html?lang=en) . Can you pls. fill an issue?

**Reply by gkzsolt** [reply]
...BUT you have to specify the chip version, because previous versions of the same board work!

**Reply by gkzsolt** [reply]
... and yes, I tried several (3) new boards which failed, while I have boards working with the previous chip.

**Reply by robert-hh** [reply]
I try to get one from this shop. Maybe they will deliver to my place and the right device.

**Reply by robert-hh** [reply]
It may take a while.

**Reply by robert-hh** [reply]
If I counted right, that will be the 21st ESP32 device of various types in my drawers,

**Reply by gkzsolt** [reply]
> I try to get one from this shop. Maybe they will deliver to my place and the right device.

This is very nice, indeed ;-) Where is your place?

> If I counted right, that will be the 21st ESP32 device of various types in my drawers,

Nice, too! I don't have so much, and I kept staying with this board lately. Previously I used ESP8266 boards, but some proved to be not so stable.

**Reply by robert-hh** [reply]
I received a board today, using the link you gave me. Unfortunately, it is not the same as yours, even if I told the shop to deliver the same one. The one I got is a standard Revision 1 model, which seems not to work as expected either: the espressif mode of switching to bootloader fails. So it's a good candidate for trash.
```
esptool.py v3.1-dev
Found 1 serial ports
Serial port /dev/ttyUSB0
Connecting........_____....._
Chip is ESP32-D0WDQ6 (revision 1)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 3c:e9:0e:4c:30:28
Uploading stub...
Running stub...
Stub running...
Warning: ESP32 has no Chip ID. Reading MAC instead.
MAC: 3c:e9:0e:4c:30:28
Hard resetting via RTS pin...
```

**Reply by cve2022** [reply]
> I received a board today, using the link you gave me. Unfortunately, it is not the same as yours, even if I told the shop to deliver the same one. The one I got is a standard Revision 1 model, which seems not to work as expected either: the espressif mode of switching to bootloader fails. So it's a good candidate for trash.

Hello @robert-hh , I did not follow the discussion, but I will hazard a guess about the problem you just described. 

Some of these ESP32 has a flaw in the hard reset system. In a nutshell, depending on the sequence of the UART's DTR and RTS, a short spurious RESET pulse is generated (I scoped once, it lasts 1 to 2us AFAIR) and that wrecks the ESP32 logic havoc (it is too short).

The uggly solution is adding a reset button to the EN pin (it works despite the switch bounce). A less inelegant solution is adding a 10uF capacitor between the EN pin and GND. The capacitor stops the spurious short reset pulse. Some boards already has this fix, from some posts I have seen.

Some links about the matter:
[https://electronics.stackexchange.com/questions/569788/how-esp32-auto-upload-issue-has-been-fixed-by-adding-capacitor](https://electronics.stackexchange.com/questions/569788/how-esp32-auto-upload-issue-has-been-fixed-by-adding-capacitor)
[https://randomnerdtutorials.com/solved-failed-to-connect-to-esp32-timed-out-waiting-for-packet-header/](https://randomnerdtutorials.com/solved-failed-to-connect-to-esp32-timed-out-waiting-for-packet-header/)

HTH

**Reply by robert-hh** [reply]
Thank you for the hint. Adding a 10µ  cap helped. I soldered it between EN and 3V3, since these pins are next to each other.

**Reply by bulletmark** [reply]
I've bought a number of ESP32 devkit clones with this problem and find it always get's solved by me soldering a cap (or adding one to my parent circuit). Here is an image I posted some time ago showing 3 clones I have bought at various times: https://imgur.com/a/YWYB2AO. Note the bottom left corner of each board. The left board has the problem, the middle board has had a cap added crudely near the EN switch, and the right board has cleaned the alignment of that cap up. I always try to ensure the board I am buying has that cap.

**Reply by cve2022** [reply]
> Thank you for the hint. Adding a 10µ cap helped. I soldered it between EN and 3V3, since these pins are next to each other.

How many pins does yours have? I am asking because mine are similar to the leftmost one at @bulletmark 's picture and the VP pin is next to EN but it is not 3V3. I mounted my ESP32 DEVKITV1 on a small protoboard and I placed the capacitor on the protoboard between EN and GND, in parallel with a button. Connecting the capacitor between EN and 3V3 must work fine too.

> I've bought a number of ESP32 devkit clones with this problem and find it always get's solved by me soldering a cap (or adding one to my parent circuit). Here is an image I posted some time ago showing 3 clones I have bought at various times: https://imgur.com/a/YWYB2AO. Note the bottom left corner of each board. The left board has the problem, the middle board has had a cap added crudely near the EN switch, and the right board has cleaned the alignment of that cap up. I always try to ensure the board I am buying has that cap.

This is the picture I remember having seen showing the different versions of the board, with a rework in the middle one and the final fix on the rightmost. Thank you for posting it, BTW. I took the liberty to reproduce it here.

![image](https://user-images.githubusercontent.com/108999978/214707443-9b80bc33-7dfe-4f1a-ace7-b5f5950597e1.png)

It is possible to solder the capacitor (I am assuming a leaded electrolytic) between the pin of the switch (like the middle board above) and GND or 3V3 (the large pin of the LDO near the GND pin on the left side of the boards. Or, over the resistor next to the EN switch ( hilighted above) between EN and 3V3.

**Comment by robert-hh** [comment]
@gkzsolt Can you make a close-up picture of your non-working device and the label of the bag in which is was shipped? I try to get the same one from the shop.

**Reply by gkzsolt** [reply]
Here you are. Sorry if it's too big. I don't have the bag, but it tells only what they call it: [NodeMCU ESP-32S dual core WIFI module](https://www.hestore.hu/prod_10037961.html?lang=en), 2x19 pins.

![ESP32-WROOM](https://user-images.githubusercontent.com/1327953/214850511-52aeda7d-6f9c-40db-9a1a-6b6bf2a32f3a.jpg)

**Reply by robert-hh** [reply]
The board I have looks very similar, if the number on the esp32 module is 211-161007. Only at my device a `K.` is printed at the antenna and the two spots at the start of the antenna are golden, not white. But that may be just the picture.
On the bottom of my module, the silkscreen reads ESP32-S, V1.1 101010. But the printing is not very clean. I doubt that this is a genuine NodeMCU device.

**Reply by gkzsolt** [reply]
> The board I have looks very similar, if the number on the esp32 module is 211-161007. 

It is.
> Only at my device a `K.` is printed at the antenna and the two spots at the start of the antenna are golden, not white. But that may be just the picture.

I cannot discern here.
> On the bottom of my module, the silkscreen reads ESP32-S, V1.1 101010. But the printing is not very clean. 

Same here, this is the bottom of the device:
![ESP32-back](https://user-images.githubusercontent.com/1327953/214860763-a3019418-ab13-4144-b118-e20ae762a3dd.jpg)

> I doubt that this is a genuine NodeMCU device.

I don't know what to say. How can you decide if it is genuine? And what is NodeMCU, btw? Wikipedia says "an open source firmware" (for which open source prototyping board designs are available). Who are the authorized board manufacturers?

Anyway, in my place this is the only ESP32 dev board available. I asked the store if they can deliver other versions, but no avail. They do not check chip versions. You can send back a board and they'll check it, but only with Arduino IDE. Micropython is not their cup of tea.

**Reply by gkzsolt** [reply]
@robert-hh , what device are you referring to (as "my device"): the one sent by Hestore? Does it finally work with the capacitor soldered?

**Reply by robert-hh** [reply]
The one sent by Hestore. At least, it works with the capacitor.

**Reply by robert-hh** [reply]
b.t.w.: Did you try another USB cable? Or a large capacitor (>1000µF) between 3V3 and GND? When transmitting starts, the device takes quite a bit of current, depending on the configured TX Power.

**Reply by gkzsolt** [reply]
> At least, it works with the capacitor.

And the Wifi works as well (you mentioned bootloader problems)?

> b.t.w.: Did you try another USB cable? Or a large capacitor (>1000µF) between 3V3 and GND? When transmitting starts, the device takes quite a bit of current, depending on the configured TX Power.

I used the same USB cable for other devices which worked. As for the power supply, I can do a test with an external supply, although I don't have much trust. More in the 10 uF capacitor ;)

**Reply by robert-hh** [reply]
Yes, WiFi works with that board.

**Reply by davefes** [reply]
> More in the 10 uF capacitor ;)

1000uF Low-ESR, if your supplier has them.

**Reply by cve2022** [reply]
> > More in the 10 uF capacitor ;)
> 
> 1000uF Low-ESR, if your supplier has them.

Not for Reset/EN, I hope.

**Reply by davefes** [reply]
Correct.  I thought he was referring to the 3V3 added cap in his quote ... my mistake.

**Reply by robert-hh** [reply]
I was referring to 3v3.

**Reply by davefes** [reply]
You were, but he had switched focus to the 10uF EN cap, which I didn't pick-up on.

**Reply by gkzsolt** [reply]
So, I put a low ESR 1000 uF cap between EN and 3V3, and... things started to shine!
(OK, I am kidding, it was only 10u:), but it really connected to WiFi. But wait.
Then I took down the cap, and the board still worked. 
(I did not solder/desolder, I used a [terminal adapter](https://www.hestore.hu/prod_10043866.html?lang=en)). This made me think.
I took the board from the terminal adapter, and left the pins "in the air" - with the USB cable still connected, of course.
The wifi connect() reset the board!!
This was my initial experiment, to try the board before soldering it on the PCB: and it failed (only starting with this chip/instance, previously never had this issue)!
So I guess soldering it in place on the PCB would work (?) Also breadboards work, with no cap.
(If somebody can come with a physical explanation, I'd be grateful;)
Well, I think I'll still solder the cap on the final circuit. But only after flashing the chip, because I could not flash it with the cap connected.
Interesting enough, isn't it? And the damn Arduino code working even "in air"...
[Note, @robert-hh  the articles indicated do not put a cap between EN and 3V3, but between EN and GND. It might still have the same effect, just an observation..]

**Reply by davefes** [reply]
Thanks for telling me you were kidding!  

> (If somebody can come with a physical explanation, I'd be grateful;)

Sounds like we are not quite ready for that ...
With a 10uF on EN to ground I have no problems flashing the board, like about 30 of them so far

**Reply by robert-hh** [reply]
That looks great. So it seems like an interference of the WiFi signal into the EN pin, which are close to each other. And maybe the terminal adapter change the capacitive load of the EN pin sufficiently to reduce the spikes or change the radiation pattern of the antenna. Arduino may use a lower TX power for sending.

About the cap between EN <-> 3V3: Assuming that the internal power source has a low impedance an the EN, it does not make a difference to EN <-> GND for a fast signal like the EN pulse. I connected the cap to 3v3, because the pins is next to EN and a 10µ SMD cap fits well there.

I wonder why you could not flash the board with the CAP connected. I could not flash it using esptool.py without the cap. I had to do the manual push-button sequence.

Edit: With a 1µ cap it works as well.

**Reply by davefes** [reply]
> So it seems like an interference of the WiFi signal into the EN pin

All my units have external antennas so hopefully that keeps most of the RF away from the PCB.  I could suggest a modification to see if that is the problem, but it would be not be easy to do.

**Reply by cve2022** [reply]
> About the cap between EN <-> 3V3: Assuming that the internal power source has a low impedance an the EN, it does not make a difference to EN <-> GND for a fast signal like the EN pulse. 

Yes to avoid the spurious reset pulse, but on second thought, connecting to ground is probably better when you connect the board to USB or a power source, as the reset active low.

**Reply by robert-hh** [reply]
It does not make a difference. The effect of the cap is to delay the trailing edge of the 100ms(!) EN pulse long enough for GPIO0 to get low. That is all related to the CP2102 resp CH340 USB/UART bridge, where on some devices the GIO0 signal seems to be somewhat delayed (or the ESP32 boots faster). Some screen shots below. Yellow is EN, green is GPIO0:
Without cap. The EN pulse ends 340µs **before** GPIO0 gets low. The duration of the EN pulse is 100ms.
![esp32_booten_wo_cap](https://user-images.githubusercontent.com/12476868/215036522-dadf04bd-8205-4dbd-b95c-4226a6ab5df7.jpg)

With 1 µ cap to 3V3. EN is at the 1.65V level 7 ms **after** GPIO0 gets low.

![esp32_booten_1u_cap](https://user-images.githubusercontent.com/12476868/215036682-07ad6924-b15f-4df8-a554-393f6ae878ab.jpg)

**Reply by davefes** [reply]
Excellent.  Curious ... in the 2nd trace the EN stays low until GPIO0 goes low, whereas in the 1st trace EN goes high before GPOI0 goes low.  I understand the ramp-up part, but there something else happening?

**Reply by robert-hh** [reply]
The espressif logic cannot pull both EN and GPIO0 low at the same time. It's either EN or GPIO0. The timing diagrams above have  different time scale. The first is 1ms/div, the second 10ms/div. So it's hard to see, whether the rising slope of the EN signal starts before of after the falling slope of GPIO0. The time difference is 340µs or about 1/30 of a div, which is "just" visible if you zoom the picture. The green falling line is right of the dashed cursor.

**Reply by robert-hh** [reply]
i have also a screen shot with a 10µ cap, which shows that the cap is just at the limit of being too large, because the EN pulse get's at the trigger level close to the end of the GPIO0 signal. So with 1µ the timing looks much better, at least for the board & set-up I have here.

![esp32_booten_10u_cap](https://user-images.githubusercontent.com/12476868/215050721-1fc3684d-f48b-446e-8927-a6527e5d849f.jpg)

**Reply by cve2022** [reply]
NOTE (EDIT): my comments below refer to [this post](https://github.com/micropython/micropython/discussions/10521#discussioncomment-4794496), I just saw Roberts's most recent post.

Interesting. Were these captured during power-on from a USB port?

That reset circuit with the two transistors have three states possible out of DTR/RTS combination to control EN/Reset and the BOOT pin GPIO0 (RTS/DTR high/high and low/low make EN/GPIO0 high). I remember in my tests with Thonny on Linuxt, to stop the running script and get back the REPL prompt, during the transition from low/low to high/high there was a brief moment where RTS was high and DTR low (or vice-versa) where this ~1us pulse appeared. I believe the ESP flasher and the Arduino programmer have the same sequence issue, thus some reports of issues with some ESP32 boards. Anyway, in this scenario (whatever programmer), the capacitor connected to either GND or 3V3 will work. I am not sure about powering-up from, say, a USB power bank or even a USB port, as the capacitor on 3V3 will pull EN up. If during the startup the CP/CH USB-Serial chip assert RTS/DTR forcing the reset, but I have no idea if that is a documented behavior (assuming it is happening). So, to me, it feel more right connecting the cap to GND as God meant it to. :smiley: .

**Reply by robert-hh** [reply]
@davefes it depends a lot on the board circuitry and the value of the pull-up resistor to 3V3. Looking at the time slope, it seems that the pull-up resistor value is ~10k. If it is lower at your board, 10µ may work fine.

@cve2022 I used esptool.py to create the time sequence, because that's the most important one for me. But the board starts fine at power-up as well. And I used the connection to 3V3 only because the EN and 3V3 pins are next to each other, and so I could solder a SMD cap between the two at the board itself.

**Reply by cve2022** [reply]
> @davefes it depends a lot on the board circuitry and the value of the pull-up resistor to 3V3. Looking at the time slope, it seems that the pull-up resistor value is ~10k. If it is lower at your board, 10µ may work fine.
> 
I believe somebody designed that small circuit with two transistors and four resistors to control the reset and boot mode (with 10k) and everybody copied that circuit. I have seen many schematics all with 10k resistors.

About using a 1uF capacitor instead of 10uF, I estimate the colector currents are lower than 100mA when the transistor connected to EN is linearly conducting (which occurs when it is discharging the capacitor). With a spurious pulse of a few us width and 100mA, the voltage drop on the 1uF cap is a few hundreds mV at most, keeping the EN pin level above the minimum needed to avoid a spurious reset of the ESP32.

**Reply by davefes** [reply]
> I believe somebody designed that small circuit with two transistors and four resistors to control the reset and boot mode (with 10k) and everybody copied that circuit.

Could you provide a link to that circuit?

**Reply by cve2022** [reply]
> > I believe somebody designed that small circuit with two transistors and four resistors to control the reset and boot mode (with 10k) and everybody copied that circuit.
> 
> Could you provide a kink to that circuit?

Absolutely. I am adding figures as who knows how long these links will be alive?
[https://oshwlab.com/Zulfahmi27/esp32-devkit-v1](https://oshwlab.com/Zulfahmi27/esp32-devkit-v1)
![image](https://user-images.githubusercontent.com/108999978/215222415-afa0d638-93bb-4ea0-8803-69c27839cd67.png)

[https://cdn-learn.adafruit.com/assets/assets/000/041/630/original/feather_schem.png?1494449413](https://cdn-learn.adafruit.com/assets/assets/000/041/630/original/feather_schem.png?1494449413)
![image](https://user-images.githubusercontent.com/108999978/215222629-105cee69-a4b4-47ba-867b-6d5231077449.png)

[https://github.com/nodemcu/nodemcu-devkit-v1.0/blob/master/NODEMCU_DEVKIT_V1.0.PDF](https://github.com/nodemcu/nodemcu-devkit-v1.0/blob/master/NODEMCU_DEVKIT_V1.0.PDF)
This one with 12k instead of 10k

**Reply by davefes** [reply]
@cve2022 thanks for those.  Good to see other implementations.

**Reply by davefes** [reply]
@cve2022  Now that I have had a good look at the "Downloader" implementation I see that my boards already have it.

My understanding is this is for uploading control.  To be honest, when I first starting using ESP32 I battled with many issues and when I read that a 1-10uF on EN was a "fix" I just incorporated it.  I could not verify that it actually fixes anything ... only that I seemed to have fewer problems.

**Reply by robert-hh** [reply]
Most schematics that I have seen use 10k. And I have built a carrier board for Pycom devices using 12k. Calculating the RC time constant, even 100nF (-> 1 ms delay) seem to be sufficient for shifting the EN rising edge into the time window of GPIO0 low.

**Reply by davefes** [reply]
All my boards have the 10K and a 100nF to ground, in which case,  my comment above should be "adding the 1uF cap probably had nothing to do with the 'flaky' behaviour I may have thought I had"!

**Reply by robert-hh** [reply]
You could test with `esptool.py --chip esp32 chip_id`. If that works fine, then the transistor logic works as expected.

b.t.w.: The google discussions logic with hierarchical responses is **irritating**; I do not like it.

**Reply by davefes** [reply]
Yes, that works.
Thanks for a way to confirm behaviour.

**Reply by cve2022** [reply]
This is the truth table of the logic:
![image](https://user-images.githubusercontent.com/108999978/215260143-04b795a7-6372-425b-8176-bd7d345ece6f.png)

This logic allows the flasher to automate the boot vs run mode. The Arduino IDE uses it and I remember seeing some articles in the internet about adding this circuit to boards that do not have it to automate the IDE debugging process.

What I observed with Thonny (on Linux) a while ago is RTS was asserted a couple of us before DTR during when I pressed the stop icon. That caused a short (2.5us IIRC) negative pulse and I believe that caused the problem as it probably violates the reset minimum width requirement (this is a supposition because Espressif's documents are terse about characteristics of the processor and I could not find info about minimum reset width in some datasheets, but it is around a tens of us (from memory) with some ST's  STM32 I worked with).
![image](https://user-images.githubusercontent.com/108999978/215261929-2d6afb1f-60a5-456a-bb36-58f6c9ef8894.png)


The 1uF/10uF capacitor filters the pulse so that the EN never reaches the reset threshold level (top trace below). I guess, with 0.1uF, the voltage drop is higher (bottom trace) and the reset threshold is reached and it wreaks havoc the microcontroller internal logic.
![image](https://user-images.githubusercontent.com/108999978/215261900-7ffeacff-822e-4daa-8859-333a73811a25.png)
That could explain why some boards may not work even with this 0.1uF capacitor. I made a quick estimation of the voltage drop with 1uF and 3us spurious pulse assuming hFE of the transistors around 300, which make the collector current 90uA and in this scenario the voltage drop is around 300mV. With 0.1uF, it would be around 3V.

The RTS before DTR assertion may be a Linux driver thing, according with the Thonny's developer, as Thonny closes and opens the port during the stop process. I am guessing it may be happening with the esptool too. The ESP32 board becoming unresponsive after a STOP happens on Thonny with Windows too, though.

I apologize if it is too much information, it may not even be the cause of the problems, but given the solution proposed in the internet (adding the capacitor) anecdotally solved the posters problems, the above could explain why it works.

**Reply by robert-hh** [reply]
I looked at the signals now with Wemos LoLin32 pro board. Somehow they manage to have both EN and GPIO0 low at the same time(!). ~~The two transistor logic has no resistors~~ It uses a two transistor package with built-in resistors of 4k7 Ohm. Screen shot below. Both boards have a short pulse on GPIO0 at the start of the sequence, like 6 µs for the NodeMCU, 2-3µs for the Wemos board. The screen shot below shows is split vertically. The upper window shows the foll EN (yellow) and GPIO0 (green) trace, the lower the Zoom on the pulse.

![esp32_booten_wemos_lolin_pro](https://user-images.githubusercontent.com/12476868/215265745-4f091659-714a-4404-9d4b-ce0f20a5541c.jpg)

**Comment by davefes** [comment]
Thank you.  I started with 1uF but ran-out of them.  I will change my 10uF to 2u2F.  Another case for "bigger is not better"!

**Reply by davefes** [reply]
@robert-hh Would you please confirm whether or not that the board that you did your testing on has the dual-transistor "Downloader" configuration?
Thanks

**Reply by robert-hh** [reply]
It has indeed the standard dual transistor logic. Sometimes boards come along which do not have that. Confusingly they have the pads for the transistors, but none soldered.

**Reply by davefes** [reply]
Good, at least we are dealing with similar hardware.

**Comment by gkzsolt** [comment]
Hi guys,

It's been a while since I posted my question and I learned that it has a workaround and perhaps an explanation. I think I will incorporate a capacitor to EN/VCC or GND, although I did not have to so far. A big thanks to all of you, especially @robert-hh, @cve2022 and @davefes. Is good to know electronics professionals worldwide.
